examples/merge.c

```diff
@@ -0,0 +1,355 @@
+		const char *curr = argv[args.pos];
+
+		if (curr[0] != '-') {
+			opts_add_refish(opts, curr);
+		} else if (!strcmp(curr, "--no-commit")) {
+			opts->no_commit = 1;
+		} else if (match_str_arg(repo_path, &args, "--git-dir")) {
+		} else {
+			print_usage();
+		}
+	}
+}
+
+int resolve_refish(git_annotated_commit **commit, git_repository *repo, const char *refish)
+{
+	git_reference *ref;
+	int err = 0;
+	git_oid oid;
+
+	assert(commit != NULL);
+
+	err = git_reference_dwim(&ref, repo, refish);
+	if (err == GIT_OK) {
+		git_annotated_commit_from_ref(commit, repo, ref);
+		git_reference_free(ref);
+		return 0;
+	}
+
+	err = git_oid_fromstr(&oid, refish);
+	if (err == GIT_OK) {
+		err = git_annotated_commit_lookup(commit, repo, &oid);
+	}
+
+	return err;
+}
+
+int resolve_heads(git_repository *repo, merge_options *opts) {
+	git_annotated_commit **annotated = calloc(opts->heads_count, sizeof(git_annotated_commit *));
+	size_t annotated_count = 0, i;
+	int err = 0;
+
+	for (i = 0; i < opts->heads_count; i++) {
+		err = resolve_refish((git_annotated_commit **)&annotated[annotated_count++], repo, opts->heads[i]);
+		if (err != 0) {
+			fprintf(stderr, "failed to resolve refish %s (%d)\n", opts->heads[i], err);
+			annotated_count--;
+			continue;
+		}
+	}
+
+	if (annotated_count <= 0) {
+		fprintf(stderr, "unable to parse any refish\n");
+		free(annotated);
+		return -1;
+	}
+
+	opts->annotated = annotated;
+	opts->annotated_count = annotated_count;
+	return 0;
+}
+
+int perform_fastforward(git_repository *repo, const git_oid *target_oid, int is_unborn)
+{
+	git_checkout_options ff_checkout_options = GIT_CHECKOUT_OPTIONS_INIT;
+	git_reference *head_ref;
+	git_reference *new_head_ref;
+	int err = 0;
+
+
+	if (is_unborn) {
+		const char *symbolic_ref;
+		git_reference *master_ref;
+
+		/* master reference is unborn, lookup HEAD manually, */
+		err = git_reference_lookup(&head_ref, repo, "HEAD");
+		if (err != 0) {
+			fprintf(stderr, "failed to lookup HEAD ref\n");
+			return -1;
+		}
+
+		/* grab its target ref, */
+		symbolic_ref = git_reference_symbolic_target(head_ref);
+
+		/* force-create it on the targets' OID, */
+		err = git_reference_create(&master_ref, repo, symbolic_ref, target_oid, 1, NULL);
+		if (err != 0) {
+			fprintf(stderr, "failed to create master reference\n");
+			return -1;
+		}
+	} else {
+		/* The repo HEAD is pointing to an existing reference. */
+		err = git_repository_head(&head_ref, repo);
+		if (err != 0) {
+			fprintf(stderr, "failed to get HEAD reference\n");
+			return -1;
+		}
+
+		/* Move the HEAD reference to the new OID... */
+		err = git_reference_set_target(&new_head_ref, head_ref, target_oid, NULL);
+		if (err != 0) {
+			fprintf(stderr, "failed to move HEAD reference\n");
+			return -1;
+		}
+	}
+
+	/* force-checkout the result. */
+	err = git_checkout_head(repo, &ff_checkout_options);
+	if (err != 0) {
+		fprintf(stderr, "failed to checkout HEAD reference\n");
+		return -1;
+	}
+	return 0;
+}
```

Comment:
I personally am not certain this is the right way to do a fast forward merge. `git_merge` can be used for this and by the end of the merge the index is set appropriately, all that remains is to update HEAD to the head of the branch you're merging. It seems you can also pass options to `git_merge` to ensure that it is fast forward only.

The other thing about this is that checkout leaves the working directory in the wrong state since it won't modify files by default.

Reply:
Well, that exact bug came up while testing 🤣. Since I'm trying to "stick to the letter", this will attempt a fast-forward if possible, and will end up with a nonsensical index (IOW the fast-forwarded commit has its reverse staged).
I'll happily fix that, I'm just obviously unable to find the correct steps (I was going by gut-feeling here : move the branch, then force-checkout).

Reply:
This looks like the correct way to do a fast-forward.  I would correct the comment - we're not _force_ checking out, just a regular checkout.